### ISSUE TITLE
Control flow analysis of streams

### ISSUE THREAD

**Issue by fprochazka** [uber/NullAway#332]
Let's consider the following example

```java
import org.javamoney.moneta.spi.DefaultNumberValue;

import javax.annotation.Nullable;
import javax.money.CurrencyUnit;
import javax.money.Monetary;
import javax.money.NumberValue;
import java.math.BigDecimal;
import java.util.*;
import java.util.stream.Collectors;

class Scratch
{

    public static void main(String[] args)
    {
        CurrencyUnit czkCurrency = Monetary.getCurrency("CZK");

        List<PriceVersion> prices = Arrays.asList(
            new PriceVersion(
                PriceType.EX_FACTORY,
                null,
                czkCurrency
            ),
            new PriceVersion(
                PriceType.EX_FACTORY,
                DefaultNumberValue.of(new BigDecimal("10")),
                czkCurrency
            )
        );

        prices.stream()
            .filter(price -> price.getValue() != null)
            .collect(Collectors.toMap(
                PriceVersion::getType,
                price -> price.getValue().numberValue(BigDecimal.class)
            ));
    }

    public static final class PriceVersion
    {

        private PriceType type;

        @Nullable
        private NumberValue value;

        private CurrencyUnit currency;

        public PriceVersion(final PriceType type, @Nullable final NumberValue value, final CurrencyUnit currency)
        {
            this.type = type;
            this.value = value;
            this.currency = currency;
        }

        public PriceType getType()
        {
            return type;
        }

        @Nullable
        public NumberValue getValue()
        {
            return value;
        }

        public CurrencyUnit getCurrency()
        {
            return currency;
        }

    }

    public enum PriceType
    {

        EX_FACTORY,

    }

}
```

the NullAway reports

> [NullAway] dereferenced expression price.getValue() is @Nullable

Which is correct, but I've previously checked that it's not and now it would be cumbersome to check again just to make "the compiler" happy.

Is basic control flow analysis of streams out of scope of this lib or would it make sense to support?

<details>
 <summary>maven: effective pom</summary>

```xml
<plugin>
    <groupId>org.apache.maven.plugins</groupId>
    <artifactId>maven-compiler-plugin</artifactId>
    <version>3.8.1</version>
    <configuration combine.self="append">
        <compilerId>javac-with-errorprone</compilerId>
        <forceJavacCompilerUse>true</forceJavacCompilerUse>
        <source>1.8</source>
        <target>1.8</target>
        <compilerArgs combine.self="override">
            <arg>-Werror</arg>
            <arg>-Xlint:all,-fallthrough,-processing,-serial,-classfile</arg>
            <arg>-parameters</arg>
            <arg>-XDcompilePolicy=simple</arg>
            <arg>-XepDisableWarningsInGeneratedCode</arg>
            <arg>-XepOpt:Immutable:KnownImmutable=javax.money.CurrencyUnit</arg>
            <arg>-Xep:ConstantField:WARN</arg>
            <arg>-Xep:ClassName:WARN</arg>
            <arg>-Xep:DivZero:WARN</arg>
            <arg>-Xep:EmptyIf:WARN</arg>
            <arg>-Xep:IterablePathParameter:WARN</arg>
            <arg>-Xep:LongLiteralLowerCaseSuffix:WARN</arg>
            <arg>-Xep:AnnotationPosition:WARN</arg>
            <arg>-Xep:EmptyTopLevelDeclaration:WARN</arg>
            <arg>-Xep:EqualsBrokenForNull:WARN</arg>
            <arg>-Xep:FunctionalInterfaceClash:WARN</arg>
            <arg>-Xep:InvalidInlineTag:WARN</arg>
            <arg>-Xep:InvalidParam:WARN</arg>
            <arg>-Xep:MissingDefault:WARN</arg>
            <arg>-Xep:NonCanonicalStaticMemberImport:WARN</arg>
            <arg>-Xep:PrimitiveArrayPassedToVarargsMethod:WARN</arg>
            <arg>-Xep:RedundantOverride:WARN</arg>
            <arg>-Xep:RedundantThrows:WARN</arg>
            <arg>-Xep:StaticQualifiedUsingExpression:WARN</arg>
            <arg>-Xep:StringEquality:WARN</arg>
            <arg>-Xep:UnusedException:WARN</arg>
            <arg>-Xep:ConstantField:WARN</arg>
            <arg>-Xep:MultiVariableDeclaration:WARN</arg>
            <arg>-Xep:MultipleTopLevelClasses:WARN</arg>
            <arg>-Xep:MultipleUnaryOperatorsInMethodCall:WARN</arg>
            <arg>-Xep:PrivateConstructorForUtilityClass:WARN</arg>
            <arg>-Xep:UngroupedOverloads:WARN</arg>
            <arg>-Xep:NullAway:ERROR</arg>
            <arg>-XepOpt:NullAway:AnnotatedPackages=com.cogvio</arg>
            <arg>-XepOpt:NullAway:TreatGeneratedAsUnannotated=true</arg>
            <arg>-XepOpt:NullAway:AcknowledgeRestrictiveAnnotations=true</arg>
            <arg>-XepOpt:NullAway:KnownInitializers=org.springframework.beans.factory.InitializingBean.afterPropertiesSet</arg>
            <arg>-XepOpt:NullAway:ExcludedFieldAnnotations=javax.persistence.PersistenceContext,org.springframework.beans.factory.annotation.Autowired</arg>
        </compilerArgs>
        <annotationProcessorPaths combine.self="append">
            <path>
                <groupId>com.uber.nullaway</groupId>
                <artifactId>nullaway</artifactId>
                <version>0.7.5</version>
            </path>
        </annotationProcessorPaths>
        <showWarnings>true</showWarnings>
    </configuration>
    <dependencies>
        <dependencies>
          <dependency>
            <groupId>org.codehaus.plexus</groupId>
            <artifactId>plexus-compiler-javac-errorprone</artifactId>
            <version>2.8.5</version>
          </dependency>
          <dependency>
            <groupId>com.google.errorprone</groupId>
            <artifactId>error_prone_core</artifactId>
            <version>2.3.3</version>
          </dependency>
          <dependency>
            <!-- this only adds our custom LibraryModels -->
            <groupId>com.cogvio.pm</groupId>
            <artifactId>build-tools-nullaway</artifactId>
            <version>0.0.1-SNAPSHOT</version>
          </dependency>
        </dependencies>
    </dependencies>
</plugin>
```

</details>

**Comment by msridhar** [comment]
We already handle this for RxJava streams.  @lazaroclapp how hard would it
be to also do it for Java 8 streams?

**Comment by lazaroclapp** [comment]
Certainly not out of scope! 

As Manu mentioned, we have some support for this already for Rx Java streams (see [RxNullabilityPropagator](https://github.com/uber/NullAway/blob/master/nullaway/src/main/java/com/uber/nullaway/handlers/RxNullabilityPropagator.java)), but not for Java 8 streams yet. This could be implemented as another handler or by extending the Rx handler into something more general.

Not exactly sure when I'll have the bandwidth to add this support, but it's certainly in the roadmap. If you are interested @fprochazka, we would be happy to accept a PR for this. It's not a trivial task, but it's also not as daunting as changing the core of NullAway would be, since all that is needed is the right `Handler` implementation. 

cc: @raghavan and @shas19 in case either of you two have some spare cycles. That said, it's not clear to me that is a priority internally just yet, though.

**Comment by fprochazka** [comment]
I can't promise anything, but I might try to send the PR (or ask my colleague to send it).

**Comment by msridhar** [comment]
I am very happy to answer questions and advise on how to get this done if
needed. Thanks, Filip!

**Comment by david-gang** [comment]
> Certainly not out of scope!
> 
> As Manu mentioned, we have some support for this already for Rx Java streams (see [RxNullabilityPropagator](https://github.com/uber/NullAway/blob/master/nullaway/src/main/java/com/uber/nullaway/handlers/RxNullabilityPropagator.java)), but not for Java 8 streams yet. This could be implemented as another handler or by extending the Rx handler into something more general.
> 
> Not exactly sure when I'll have the bandwidth to add this support, but it's certainly in the roadmap. If you are interested @fprochazka, we would be happy to accept a PR for this. It's not a trivial task, but it's also not as daunting as changing the core of NullAway would be, since all that is needed is the right `Handler` implementation.
> 
> cc: @raghavan and @shas19 in case either of you two have some spare cycles. That said, it's not clear to me that is a priority internally just yet, though.

Hi @lazaroclapp ,
I read the documentation of handler

> 
> /**
>  * The general interface representing a handler.
>  *
>  * <p>Handlers are used to model specific libraries and APIs, as opposed to general features of the
>  * Java language, which are handled by the nullability checker core and dataflow packages.
>  */

Isn't stream a general feature of the java language? So is implementing it with a Handler the best way ?

**Comment by lazaroclapp** [comment]
> Hi @lazaroclapp ,
> I read the documentation of handler
> 
> > /**
> > 
> > * The general interface representing a handler.
> > 
> > * Handlers are used to model specific libraries and APIs, as opposed to general features of the
> > * Java language, which are handled by the nullability checker core and dataflow packages.
> >   */
> 
> Isn't stream a general feature of the java language? So is implementing it with a Handler the best way ?

You could argue this one either way, because while streams are part of the standard JRE library, they are not language features in the sense of having specialized syntax (they are just method calls with some conventions that take advantage of lambdas which _are_ a core language feature). Stream APIs are not mentioned once in the JLS. 

That said, the real reason is: we have handlers support for RxJava, which is a clearly external library, and the support for Java streams is mostly analogous and uses the same hook points, so we can do it without changes to the core. I also would prefer not to do it inside the core, because it would be a one-off case where the analysis is inter-procedural instead of purely intra-procedural, and that seems like a bad fit for core (i.e. it should be possible to disable all handlers and get a predictable if less practical nullness checker out of NullAway, introduce stream handling in core and that basic nullness checker becomes "Intra-procedural, except when..."). 

I don't know how principled those arguments really are, though. But my feeling is that this code will be cleaner outside of core.

cc: @msridhar thoughts?

**Comment by msridhar** [comment]
I agree that Java 8 stream support should be in a handler for the reasons that @lazaroclapp outlines.

**Comment by david-gang** [comment]
i started to work on this.

**Comment by msridhar** [comment]
Let's assume this was closed by #371 and open new issues for further support of streams
